### PR TITLE
Update Test Application (npm run, azure update, fix deprecated usage)

### PR DESCRIPTION
**Pull request by bobvanmanen**
Add `npm run testbed` to run the test application
Remove nested package.json
Upgrade azure storage sdk
Fixed deprecated URL.parse

### CODE REVIEW

**Comment by bobvanmanen** [comment]
Includes changes in https://github.com/adobe/node-httptransfer/pull/37 which should be merged first.

File: lib/headers.js [hunk 1]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+ * Last modified header in milliseconds since epoch
+ * 
+ * @param {Headers} headers HTTP headers
+ * @returns {number} Last-modified time stamp in milliseconds since epoch
+ */
+function getLastModified(headers) {
+    const lastModified = Date.parse(headers.get(HTTP.HEADER.LAST_MODIFIED));
+    if (!isNaN(lastModified)) {
+        return lastModified;
+    }
```
Comment (mfrisbey) [inline lib/headers.js:49]: Does this need a return value if `isNan()` is true?

File: lib/headers.js [hunk 2]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+ * Non-weak ETag identifier
+ * 
+ * @param {Headers} headers HTTP headers
+ * @returns {string} Non-weak ETag identifier
+ */
+function getETag(headers) {
+    const etag = headers.get(HTTP.HEADER.ETAG);
+    if (etag && !etag.startsWith("W/")) {
+        return etag;
+    }
```
Comment (mfrisbey) [inline lib/headers.js:62]: Similar here - does it need a return value if the condition isn't met?

File: lib/headers.js [hunk 3]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+ * Retrieve and parse Content-Range from the HTTP headers
+ * 
+ * @param {Headers} headers HTTP headers
+ * @returns {ContentRange} Content range
+ */
+function getContentRange(headers) {
+    const contentRange = headers.get(HTTP.HEADER.CONTENT_RANGE);
+    if (contentRange) {
+        return parseContentRange(contentRange);
+    }
```
Comment (mfrisbey) [inline lib/headers.js:151]: Another missing return value?

File: lib/headers.js [hunk 4]
```diff
@@ -27,49 +28,155 @@ async function getHeaders(url, options={}) {
...
+function getContentLength(headers) {
+    const value = headers.get(HTTP.HEADER.CONTENT_LENGTH);
+    if (value) {
+        const parsed = parseInt(value);
+        if (typeof parsed === "number" && !isNaN(parsed)) {
+            return parsed;
         }
-    } else {
-        return null;
     }
```
Comment (mfrisbey) [inline lib/headers.js:167]: Ditto.

**Review by mfrisbey — CHANGES_REQUESTED** [review]
Looks great - just a few potential missing return values?

**Comment by bobvanmanen** [comment]
> Looks great - just a few potential missing return values?

They're not strictly necessary, but without the return values it is a bit confusing. These changes were part of an earlier PR. I've opened a PR to fix it though: https://github.com/adobe/node-httptransfer/pull/41

**Review by mfrisbey — APPROVED** [review]
Thanks!

File: testbed/index.js [hunk 5]
```diff
@@ -14,77 +14,86 @@
...
+        sharedKeyCredential
     );
+    return blobServiceClient.getContainerClient(containerName);
+}
+
+function createAzureSAS(auth, containerName, blobName, perm="r") {
+    const containerClient = createAzureContainerClient(auth, containerName);
+
+    const permissions = new BlobSASPermissions();
+    permissions.read = (perm.indexOf("r") >= 0);
```
Comment (tmathern) [inline testbed/index.js:53]: If `perm` is a string, `includes` would work too.

File: testbed/index.js [hunk 6]
```diff
@@ -14,77 +14,86 @@
...
+    const containerClient = createAzureContainerClient(auth, containerName);
+    const blobClient = containerClient.getBlockBlobClient(blobName);
+    const blockList = await blobClient.getBlockList("uncommitted");
+    await blobClient.commitBlockList(blockList.uncommittedBlocks.map(x => x.name));
 }
 
 async function resolveLocation(value, options) {
-    const url = URL.parse(value);
-    if (url.protocol === "https:" || url.protocol === "http:") {
+    if (value.startsWith("https:") || value.startsWith("http:")) {
```
Comment (tmathern) [inline testbed/index.js:79]: Usually we don't allow `http`. We require `https`...

**Review by tmathern — COMMENTED** [review]
LGTM, with some comments

File: testbed/index.js (hunk 6 above)
Comment (bobvanmanen) [reply testbed/index.js:79]: I agree, but this is a test application and I'd like to be able to upload to localhost.

**Pull request merged**